File: src/rate_of_closure/variation/_localized_attribution_contract.py
"""Shared bounds and numerical policy for localized attribution authority."""

from __future__ import annotations

import math
import sys
from dataclasses import dataclass

from shared.python.contracts import require

MAX_TEXT_LENGTH = 256
MAX_SAFE_INTEGER = 9_007_199_254_740_991
MAX_SOURCES = 32
MAX_TARGETS = 64
MAX_PAIRS = 4096
MAX_OBSERVATIONS = 131_072
RESPONSE_ULPS = 4.0
STATUS_VALUES = frozenset({"evaluated_hit", "evaluated_no_impact", "numerical_failure"})


@dataclass(frozen=True)
class TargetDefinition:
    """Canonical meaning for one target name."""

    kind: str
    unit: str
    convention: str
    coordinate_frame: str | None


TARGET_REGISTRY = {
    "position_x_m": TargetDefinition(
        "state", "m", "app-frame-cartesian-v1", "app_frame:x_target,y_up,z_right"
    ),
    "position_y_m": TargetDefinition(
        "state", "m", "app-frame-cartesian-v1", "app_frame:x_target,y_up,z_right"
    ),
    "position_z_m": TargetDefinition(
        "state", "m", "app-frame-cartesian-v1", "app_frame:x_target,y_up,z_right"
    ),
    "impact_time_s": TargetDefinition("impact", "s", "rate-of-closure-impact-v1", None),
    "clubhead_speed_mps": TargetDefinition(
        "impact", "m/s", "rate-of-closure-impact-v1", None
    ),
    "spin_loft_deg": TargetDefinition(
        "impact", "deg", "rate-of-closure-impact-v1", None
    ),
    "face_to_path_deg": TargetDefinition(
        "impact", "deg", "rate-of-closure-impact-v1", None
    ),
    "spin_axis_tilt_deg": TargetDefinition(
        "impact", "deg", "rate-of-closure-impact-v1", None
    ),
    "ball_speed_mph": TargetDefinition(
        "shot", "mph", "rate-of-closure-flight-v1", None
    ),
    "launch_angle_deg": TargetDefinition(
        "shot", "deg", "rate-of-closure-flight-v1", None
    ),
    "launch_azimuth_deg": TargetDefinition(
        "shot", "deg", "rate-of-closure-flight-v1", None
    ),
    "spin_rpm": TargetDefinition("shot", "rpm", "rate-of-closure-flight-v1", None),
    "carry_m": TargetDefinition("shot", "m", "rate-of-closure-flight-v1", None),
    "lateral_m": TargetDefinition("shot", "m", "rate-of-closure-flight-v1", None),
    "max_height_m": TargetDefinition("shot", "m", "rate-of-closure-flight-v1", None),
    "flight_time_s": TargetDefinition("shot", "s", "rate-of-closure-flight-v1", None),
    "landing_angle_deg": TargetDefinition(
        "shot", "deg", "rate-of-closure-flight-v1", None
    ),
}


def require_bounded_count(size: int, maximum: int, label: str) -> None:
    """Reject resource-heavy arrays before dependent allocation."""
    require(0 <= size <= maximum, f"{label} exceeds resource cap", (size, maximum))


def require_authority_shape(
    sources: int, targets: int, pairs: int, observations: int
) -> None:
    """Bound all arrays and the complete matrix before materialization."""
    require_bounded_count(sources, MAX_SOURCES, "sources")
    require_bounded_count(targets, MAX_TARGETS, "targets")
    require_bounded_count(pairs, MAX_PAIRS, "pairs")
    require_bounded_count(observations, MAX_OBSERVATIONS, "observations")
    require(
        pairs * targets <= MAX_OBSERVATIONS, "pair-target matrix exceeds resource cap"
    )


def response_matches(actual: float, expected: float) -> bool:
    """Use the cross-runtime four-scaled-ULP response policy."""
    tolerance = RESPONSE_ULPS * sys.float_info.epsilon * max(1.0, abs(expected))
    return math.isfinite(actual) and abs(actual - expected) <= tolerance


__all__ = [
    "MAX_OBSERVATIONS",
    "MAX_PAIRS",
    "MAX_SAFE_INTEGER",
    "MAX_SOURCES",
    "MAX_TARGETS",
    "MAX_TEXT_LENGTH",
    "TARGET_REGISTRY",
    "require_bounded_count",
    "require_authority_shape",
    "response_matches",
]

File: src/rate_of_closure/web/src/model/localizedAttributionContract.ts
/** Shared strict bounds, target meanings, and numerical policy. */

export type RecordValue = Record<string, unknown>;

export const MAX_TEXT_LENGTH = 256;
export const MAX_SOURCES = 32;
export const MAX_TARGETS = 64;
export const MAX_PAIRS = 4096;
export const MAX_OBSERVATIONS = 131072;

interface TargetDefinition {
  kind: "state" | "impact" | "shot";
  unit: string;
  convention: string;
  coordinateFrame: string | null;
}

const state = (unit = "m"): TargetDefinition => ({
  kind: "state", unit, convention: "app-frame-cartesian-v1",
  coordinateFrame: "app_frame:x_target,y_up,z_right",
});
const impact = (unit: string): TargetDefinition => ({
  kind: "impact", unit, convention: "rate-of-closure-impact-v1",
  coordinateFrame: null,
});
const shot = (unit: string): TargetDefinition => ({
  kind: "shot", unit, convention: "rate-of-closure-flight-v1",
  coordinateFrame: null,
});

export const TARGET_REGISTRY: Readonly<Record<string, TargetDefinition>> = {
  position_x_m: state(), position_y_m: state(), position_z_m: state(),
  impact_time_s: impact("s"), clubhead_speed_mps: impact("m/s"),
  spin_loft_deg: impact("deg"), face_to_path_deg: impact("deg"),
  spin_axis_tilt_deg: impact("deg"), ball_speed_mph: shot("mph"),
  launch_angle_deg: shot("deg"), launch_azimuth_deg: shot("deg"),
  spin_rpm: shot("rpm"), carry_m: shot("m"), lateral_m: shot("m"),
  max_height_m: shot("m"), flight_time_s: shot("s"),
  landing_angle_deg: shot("deg"),
};

export const record = (
  value: unknown, fields: readonly string[], label: string,
): RecordValue => {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${label} must be an object`);
  }
  const result = value as RecordValue;
  if (Object.keys(result).sort().join("|") !== [...fields].sort().join("|")) {
    throw new Error(`${label} has invalid fields`);
  }
  return result;
};

export const stable = (value: unknown, label: string): string => {
  const hasControl = typeof value === "string" && [...value].some(
    (character) => character.charCodeAt(0) < 32,
  );
  if (typeof value !== "string" || value.length === 0 || value.trim() !== value ||
      [...value].length > MAX_TEXT_LENGTH || hasControl || /^[=+\-@]/u.test(value)) {
    throw new Error(`${label} must be a stable safe ID within the length cap`);
  }
  return value;
};

export const finite = (value: unknown, label: string): number => {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error(`${label} must be finite`);
  }
  return value;
};

export const nullableFinite = (value: unknown, label: string): number | null =>
  value === null ? null : finite(value, label);

export const index = (value: unknown, label: string): number => {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0) {
    throw new Error(`${label} must be a nonnegative safe integer`);
  }
  return value;
};

export const boundedArray = (
  value: unknown, maximum: number, label: string,
): readonly unknown[] => {
  if (!Array.isArray(value) || value.length > maximum) {
    throw new Error(`${label} must be an array within its resource cap`);
  }
  return value;
};

export const requireAuthorityShape = (
  sources: number, targets: number, pairs: number, observations: number,
): void => {
  if (sources > MAX_SOURCES || targets > MAX_TARGETS || pairs > MAX_PAIRS ||
      observations > MAX_OBSERVATIONS || pairs * targets > MAX_OBSERVATIONS) {
    throw new Error("authority arrays or pair-target matrix exceed resource cap");
  }
};

export const responseMatches = (actual: number, expected: number): boolean => {
  const tolerance = 4 * Number.EPSILON * Math.max(1, Math.abs(expected));
  return Number.isFinite(actual) && Number.isFinite(expected) &&
    Number.isFinite(tolerance) && Math.abs(actual - expected) <= tolerance;
};

export const requireNonzeroIntervention = (baseline: number, perturbed: number): void => {
  if (baseline === perturbed) throw new Error("source intervention delta must be nonzero");
};

export const deepFreeze = <T>(value: T): T => {
  if (typeof value === "object" && value !== null && !Object.isFrozen(value)) {
    Object.values(value as Record<string, unknown>).forEach(deepFreeze);
    Object.freeze(value);
  }
  return value;
};
